### PR TITLE
Fixes #57

### LINKED ISSUE #57: S3 bucket cannot be ref://
When trying to do 
```
s3:
  upload_lambda_trigger:
    bucket: ref://InputBucket
```
Gordon fails with 
```
Building project...
  0001_p.json
      cp -Rf * /Users/tcosta/.gordon/tmpIxTWCd
      pip install --install-option="--prefix=" -r requirements.txt -q -t /Users/tcosta/.gordon/tmpIxTWCd
      /Library/Python/2.7/site-packages/pip-8.1.2-py2.7.egg/pip/commands/install.py:180: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
        cmdoptions.check_install_build_global(options)

      cd /Users/tcosta/.gordon/tmpIxTWCd && find . -name "*.pyc" -delete
      cp -Rf * /Users/tcosta/.gordon/tmpWuqh_1
      pip install --install-option="--prefix=" -r requirements.txt -q -t /Users/tcosta/.gordon/tmpWuqh_1
      /Library/Python/2.7/site-packages/pip-8.1.2-py2.7.egg/pip/commands/install.py:180: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
        cmdoptions.check_install_build_global(options)

      cd /Users/tcosta/.gordon/tmpWuqh_1 && find . -name "*.pyc" -delete
      cp -Rf * /Users/tcosta/.gordon/tmpViyHe0
      cp -Rf * /Users/tcosta/.gordon/tmpJovOg2
  0002_pr_r.json
Traceback (most recent call last):
  File "/usr/local/bin/gordon", line 9, in <module>
    load_entry_point('gordon==0.2.1', 'console_scripts', 'gordon')()
  File "/Library/Python/2.7/site-packages/gordon/bin.py", line 106, in main
    getattr(obj, options.func)()
  File "/Library/Python/2.7/site-packages/gordon/core.py", line 243, in build
    self._build_resources_template()
  File "/Library/Python/2.7/site-packages/gordon/core.py", line 334, in _build_resources_template
    r.register_resources_template(template)
  File "/Library/Python/2.7/site-packages/gordon/resources/s3.py", line 337, in register_resources_template
    Bucket=self.get_bucket_name(),
  File "/Library/Python/2.7/site-packages/gordon/resources/s3.py", line 274, in get_bucket_name
    if bucket.startswith('arn:aws:'):
AttributeError: 'Ref' object has no attribute 'startswith'
```

This is with debug mode on. The ref is defined in `dev.yml` as
```
---
InputBucket: my-test-bucket
```

The build succeeds if the ref is used in the context property and the ref is removed from the s3 key.

### PR DESCRIPTION
**Pull request by tjsail33**
This PR addressed the broken bucket refs in gordon.contribs.s3 for buckets mentioned in #57

### CODE REVIEW

**Comment by camfraser** [comment]
👍

**Comment by bsiemon** [comment]
Is there a version of gordon not impacted by this bug?

**Comment by tjsail33** [comment]
Not that I have seen. We have a project that is delayed from going to production due to this issue. I'm hoping @jorgebastida can merge this in and tag in the next day or two.

**Comment by bsiemon** [comment]
Is it worth fixing https://github.com/jorgebastida/gordon/issues/34 in this PR also? Seems like there is a systemic bug around treating values that could be `refs` as strings in the code.

File: gordon/resources/s3.py [hunk 1]
```diff
@@ -265,13 +265,13 @@ def __init__(self, *args, **kwargs):
 
     def get_bucket_arn(self):
         bucket = self.settings.get('bucket')
-        if bucket.startswith('arn:aws:'):
+        if isinstance(bucket, str) and bucket.startswith('arn:aws:'):
```
Comment (bsiemon) [inline gordon/resources/s3.py]: Other spots in s3.py seems to prefer `isinstance(destination, six.string_types):` over just checking `str`. 

https://github.com/jorgebastida/gordon/blob/master/gordon/resources/s3.py#L127

**Comment by bsiemon** [comment]
https://github.com/jorgebastida/gordon/pull/62 is in the same vein as this fix.

File: gordon/resources/s3.py [hunk 2]
```diff
@@ -265,13 +265,13 @@ def __init__(self, *args, **kwargs):
 
     def get_bucket_arn(self):
         bucket = self.settings.get('bucket')
-        if bucket.startswith('arn:aws:'):
+        if isinstance(bucket, six.string_types) and bucket.startswith('arn:aws:'):
             return bucket
         return troposphere.Join("", ["arn:aws:s3:::", bucket])
```
Comment (bsiemon) [inline gordon/resources/s3.py]: What will happen when bucket is a Ref and it is then joined with a string?

This method might need to detect it is dealing with a Ref then just return the Ref instance unchanged.

Comment (camfraser) [inline gordon/resources/s3.py]: @bsiemon you should probably checkout the troposphere repo. It has some examples of using `troposphere.Join` for exactly that purpose: https://github.com/cloudtools/troposphere/blob/master/examples/Route53_RoundRobin.py#L29

**Comment by tjsail33** [comment]
@jorgebastida any news on when this PR will be merged? It is actively blocking a production deployment of a new tool.

**Comment by tjsail33** [comment]
@jorgebastida is this project abandoned? Should I fork it and release a `gordon-alt` or something like that on PyPi until  you are back? It's been a month and a half and the lack of merging PRs is actively blocking a production deployment, so some communication would be appreciated.

**Comment by jorgebastida** [comment]
This project is not abandoned, no. Have had a couple of rough months in which I had better things to do than merging PR I'm afraid :/

That been said, I can't see why merging a PR in an open-source project might be blocking anything! Feel free to create your fork, merge anything you might need and move from there.

Hopefully, I will start catching up soon, but can't promise anything, sorry.

**Comment by tjsail33** [comment]
I suppose you and I have different definitions of "maintaining" when it comes to OSS then. Apologies.

**Comment by bsiemon** [comment]
I think this project solves the lambda config problem in a great way. Use of cloud formation sets it apart from various other solutions. 

@jorgebastida are you interested in help maintaining this project going forward? I think projects like this will only become more popular in the future since lambda provides a great way to glue bits of an application together.

**Comment by camfraser** [comment]
I would also be willing to help with maintaining this project going forward.

**Comment by tjsail33** [comment]
I am also very interested and willing to assist in maintaining Gordon.

**Comment by tjsail33** [comment]
@bsiemon I have published `gordon-alt` at https://pypi.python.org/pypi/gordon-alt/0.2.2. Feel free to submit any PRs against master of that repo and I will merge them in.

**Comment by bsiemon** [comment]
Thanks for doing the leg work to set that up. I can recreate my existing PR pretty easy.

**Comment by dmelamedcl** [comment]
+1 for helping maintaining this project!

**Comment by jorgebastida** [comment]
Just pushed https://github.com/jorgebastida/gordon/pull/77 which fixes #57 
Will try to put together some other bugfixes and ship ``0.3.0`` before the end of the day.

Thank you very much for your interest in maintaining the project  @bsiemon @camfraser @tjsail33 @dmelamedcl I am quite certaily looking for help.

Some time ago I did setup a gitter channel for the project (https://gitter.im/jorgebastida/gordon) would you be interesting in joining so we can coordinate better?

**Pull request closed without merging**